{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyPuZ3c8lF6n4gPXTyrc3nvt"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","source":["import pandas as pd\n","from itertools import product\n","import openpyxl\n","from openpyxl.styles import PatternFill\n","import random\n","from collections import defaultdict"],"metadata":{"id":"PMnaOSDW5eju"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["file_name = 'punnett_square.xlsx'"],"metadata":{"id":"oDY45SUN5Py8"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# input the number of crosses\n","num_crosses = int(input(\"Please enter the number of crosses: \"))\n","\n","# generate the genes assuming two alleles per gene (e.g., Aa, Bb)\n","genes = [chr(65+i)+chr(97+i) for i in range(num_crosses)]\n","\n","# generate all possible combinations of alleles\n","gametes = [''.join(gamete) for gamete in product(*genes)]\n","\n","# initialize an empty DataFrame\n","df = pd.DataFrame(index=gametes, columns=gametes)\n","\n","def sort_characters(text):\n","    # sort the characters alphabetically, then by uppercase-lowercase within each alphabet group\n","    sorted_characters = sorted(text, key=lambda char: (char.lower(), not char.isupper()))\n","\n","    # join the sorted characters back into a string\n","    return ''.join(sorted_characters)\n","\n","# fill in the DataFrame with combinations of gametes\n","genotypes = []\n","for i in gametes:\n","    for j in gametes:\n","        # combine and sort each pair of alleles, then join them back into strings\n","        gamete = sort_characters(''.join(sorted(i[k]+j[k] for k in range(len(i)))))\n","        genotypes.append(gamete)\n","        df.loc[i,j] = gamete\n","\n","# write to Excel file\n","df.to_excel(file_name)\n","\n","# done!\n","print(f'done with {len(genotypes)}(2^{2*num_crosses}) genotypes!')\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Yt1fmAGWsCaq","executionInfo":{"status":"ok","timestamp":1693492010085,"user_tz":-540,"elapsed":4493,"user":{"displayName":"김규민","userId":"07021340244099797945"}},"outputId":"6138b2ff-adee-41f5-e9f3-b733b1e79dd3"},"execution_count":66,"outputs":[{"output_type":"stream","name":"stdout","text":["Please enter the number of crosses: 2\n","done with 16(2^4) genotypes!\n"]}]},{"cell_type":"code","source":["def generate_random_color():\n","  r = lambda: random.randint(128,255)\n","  return '{:02X}{:02X}{:02X}'.format(r(),r(),r())\n","\n","# load workbook and select active sheet\n","wb = openpyxl.load_workbook(file_name)\n","ws = wb.active\n","\n","# create a dictionary to store cell values and their corresponding colors\n","cell_dict = {}\n","\n","for row in ws.iter_rows(min_row=2):\n","    for cell in row:\n","        if cell.column == 1:\n","            continue\n","\n","        if cell.value not in cell_dict:\n","            # generate a new color for this unique value\n","            color_code = generate_random_color()\n","            fill_color = PatternFill(start_color=color_code,\n","                                    end_color=color_code,\n","                                    fill_type=\"solid\")\n","            cell_dict[cell.value] = fill_color\n","\n","        # fill the cell with its corresponding color\n","        cell.fill = cell_dict[cell.value]\n","\n","# write to new Excel file\n","wb.save('colored_' + file_name)\n","\n","# done!\n","print(f'done with beautiful {len(cell_dict)} colors!')\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Og39kw5aSxHo","executionInfo":{"status":"ok","timestamp":1693492013009,"user_tz":-540,"elapsed":429,"user":{"displayName":"김규민","userId":"07021340244099797945"}},"outputId":"c785fb85-5193-492a-e1ad-b4375411426a"},"execution_count":67,"outputs":[{"output_type":"stream","name":"stdout","text":["done with beautiful 9 colors!\n"]}]},{"cell_type":"code","source":["# initialize the phenotype counter\n","phenotypes = defaultdict(int)\n","\n","# count phenotypes\n","for genotype in genotypes:\n","    phenotype = ''\n","\n","    for i in range(0, len(genotype), 2):\n","        if genotype[i].isupper():\n","            trait = genotypes[0][2*(i//2+1)-1]\n","        else:\n","            trait = (genotypes[0][2*(i//2+1)-1]).lower()\n","\n","        phenotype += trait\n","\n","    phenotypes[phenotype] += 1\n","\n","print('============================')\n","print('    < phenotypic ratio >    ')\n","print('* uppercase: dominant  trait')\n","print('* lowercase: recessive trait')\n","print('============================')\n","\n","# sort and print results in descending order\n","sorted_result = sorted(phenotypes.items(), key=lambda x: x[1], reverse=True)\n","for k, v in sorted_result:\n","     print(f\"{k}: {v}\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"BBNMNSCEpEbw","executionInfo":{"status":"ok","timestamp":1693492015776,"user_tz":-540,"elapsed":5,"user":{"displayName":"김규민","userId":"07021340244099797945"}},"outputId":"453cd5c6-16c0-4cda-814d-351c521b76d4"},"execution_count":68,"outputs":[{"output_type":"stream","name":"stdout","text":["============================\n","    < phenotypic ratio >    \n","* uppercase: dominant  trait\n","* lowercase: recessive trait\n","============================\n","AB: 9\n","Ab: 3\n","aB: 3\n","ab: 1\n"]}]}]}